pyuvdata/calfits.py

```diff
@@ -0,0 +1,194 @@
+            prihdr['CTYPE1'] = ('Narrays', 'Number of image arrays.')
+            prihdr['CUNIT1'] = ('Integer', 'Number of image arrays. Increment.')
+            prihdr['CRVAL1'] = (3, 'Number of image arryays.')
+            prihdr['CDELT1'] = 1
+
+            pridata = np.concatenate([self.gain_array.real[:, :, :, :, np.newaxis],
+                                      self.gain_array.imag[:, :, :, :, np.newaxis],
+                                      # self.flag_array[:, :, :, :, np.newaxis],
+                                      self.quality_array[:, :, :, :, np.newaxis]],
+                                     axis=-1)
+
+        if self.cal_type == 'delay':
+            # Set header variable for gain.
+            prihdr['CTYPE4'] = ('FREQS', 'Valid frequencies to apply delay.')
+            prihdr['CUNIT4'] = ('Hz', 'Units of frequecy.')
+            prihdr['CRVAL4'] = self.freq_array[0][0]
+            prihdr['CDELT4'] = self.channel_width
+            # set the last axis for number of arrays.
+            prihdr['CTYPE1'] = ('Narrays', 'Number of image arrays.')
+            prihdr['CUNIT1'] = ('Integer', 'Number of image arrays. Value.')
+            prihdr['CRVAL1'] = (2, 'Number of image arrays.')
+            prihdr['CDELT1'] = 1
+
+            pridata = np.concatenate([self.delay_array[:, :, :, :, np.newaxis],
+                                      # self.flag_array[:, :, :, :, np.newaxis],
+                                      self.quality_array[:, :, :, :, np.newaxis]],
+                                     axis=-1)
+
+        if self.cal_type == 'unknown':
+            raise ValueError("unknown calibration type. Do not know how to"
+                             "store parameters")
+
+        # header ctypes for NAXIS
+        prihdr['CTYPE5'] = ('ANTS', 'Antenna numbering.')
+        prihdr['CUNIT5'] = 'Integer'
+        prihdr['CRVAL5'] = 0
+        prihdr['CDELT5'] = -1
+
+        prihdr['CTYPE3'] = ('TIME', 'Time axis.')
+        prihdr['CUNIT3'] = ('JD', 'Time in julian date format')
+        prihdr['CRVAL3'] = self.time_array[0]
+        prihdr['CDELT3'] = self.integration_time
+
+        # more checks for polarization. check ordering.
+        prihdr['CTYPE2'] = ('JONES', 'Jones matrix array')
+        prihdr['CUNIT2'] = ('Integer', 'representative integer for polarization.')
+        prihdr['CRVAL2'] = self.jones_array[0]  # always start with first jones.
+        prihdr['CDELT2'] = -1
+
+        prihdu = fits.PrimaryHDU(data=pridata, header=prihdr)
+
+        col1 = fits.Column(name='ANTNAME', format='8A',
+                           array=self.antenna_names)
+        col2 = fits.Column(name='ANTINDEX', format='D',
+                           array=self.antenna_numbers)
+        cols = fits.ColDefs([col1, col2])
+        ant_hdu = fits.BinTableHDU.from_columns(cols)
+
+        sechdr = fits.Header()
+        secdata = self.flag_array.astype(np.int64) # Can't be bool
```

Comment:
It doesn't look like you have any axis descriptions for this secondary header which contains the flags. I don't think there's enough information without it is there? Also, when you have a gain cal type, I think the flags should go into the same HDU with the data to cut down on confusion (and so you don't have to repeat all the header stuff). I know that will lead to variable numbers of columns for the different types, but I think that's ok (it certainly happens in uvfits files). I tend to think that the gain type is going to be the most common file type (since you need it to represent a full calibration solution), so I think we want it to optimize for it.

I also don't see anything either here in the write or in the read about the input_flag_array. I think we should write it and read it if it exists.